contracts/distribution/subdomains/ZNSSubdomainRegistrar.sol

```diff
@@ -0,0 +1,229 @@
+            || accessController.isRegistrar(msg.sender),
+            "ZNSSubdomainRegistrar: Not authorized"
+        );
+        _;
+    }
+
+    // TODO sub: proxy ??
+    constructor(
+        address _accessController,
+        address _registry,
+        address _rootRegistrar
+    ) {
+        _setAccessController(_accessController);
+        setRegistry(_registry);
+        setRootRegistrar(_rootRegistrar);
+    }
+
+    function registerSubdomain(
+        bytes32 parentHash,
+        string calldata label,
+        address domainAddress,
+        DistributionConfig calldata distrConfig
+    ) external override returns (bytes32) {
+        // TODO sub: make the order of ops better
+        DistributionConfig memory parentConfig = distrConfigs[parentHash];
+        require(
+            parentConfig.accessType != AccessType.LOCKED
+                || registry.isOwnerOrOperator(parentHash, msg.sender),
+            "ZNSSubdomainRegistrar: Parent domain's distribution is locked"
+        );
+
+        if (parentConfig.accessType == AccessType.WHITELIST) {
+            require(
+                distributionWhitelist[parentHash][msg.sender],
+                "ZNSSubdomainRegistrar: Sender is not whitelisted"
+            );
+        }
+
+        bytes32 subdomainHash = hashWithParent(parentHash, label);
+
+        require(
+            !registry.exists(subdomainHash),
+            "ZNSSubdomainRegistrar: Subdomain already exists"
+        );
+
+        uint256 price;
+        uint256 fee;
+        // TODO sub: can we make this abstract switching better ??
+        if (parentConfig.pricingContract.feeEnforced()) {
+            (price, fee) = AZNSPricingWithFee(address(parentConfig.pricingContract)).getPriceAndFee(
+                parentHash,
+                label
+            );
+        } else {
+            price = parentConfig.pricingContract.getPrice(parentHash, label);
+        }
+
+        parentConfig.paymentContract.processPayment(
+            parentHash,
+            subdomainHash,
+            msg.sender,
+            price,
+            fee
+        );
+
+        rootRegistrar.coreRegister(
+            parentHash,
+            subdomainHash,
+            label,
+            msg.sender,
+            domainAddress
+        );
+
+        if (address(distrConfig.pricingContract) != address(0)
+            && address(distrConfig.paymentContract) != address(0)) {
+            setDistributionConfigForDomain(subdomainHash, distrConfig);
+        }
+
+        return subdomainHash;
+    }
+
+    function revokeSubdomain(bytes32 parentHash, bytes32 domainHash) external override {
+        // TODO sub: can this be combined with the same check in the Main Registrar ??
+        require(
+            rootRegistrar.isOwnerOf(domainHash, msg.sender, IZNSRegistrar.OwnerOf.BOTH),
+            "ZNSSubdomainRegistrar: Not the owner of both Name and Token"
+        );
+
+        rootRegistrar.coreRevoke(domainHash);
```

Comment:
`coreRevoke()` wipes the main domain data only, which is in registry and domain token. It also now sets whatever price was set by the revokee to 0, so that new registrations do not have to pay to the guy who has no domain anymore.
refunds so far are only issued by StakePayment, which keeps it's own state of how much each person staked, so that wiping the price doesn't do anything.

Reply:
and yes, there's already a ton of tests for refunds :)